Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp

```diff
@@ -0,0 +1,263 @@
+        for (int i = 0; i < wave->num_samples; i++)
+            data[i * priv->info.channels] = wave->samples[i];
+
+        gst_adapter_push(members->adapter.get(), gst_buffer_ref(buf.get()));
+    }
+
+    size_t queueSize = gst_adapter_available(members->adapter.get());
+    if (members->didLoadUtterance && !queueSize)
+        return GST_FLOW_EOS;
+
+    *buffer = gst_adapter_take_buffer(members->adapter.get(), std::min(queueSize, bytes));
+    return GST_FLOW_OK;
+}
+
+static gboolean webkitFliteSrcSetCaps(GstBaseSrc* baseSource, GstCaps* caps)
+{
+    WebKitFliteSrc* src = WEBKIT_FLITE_SRC(baseSource);
+    WebKitFliteSrcPrivate* priv = src->priv;
+
+    gst_audio_info_init(&priv->info);
+    if (!gst_audio_info_from_caps(&priv->info, caps)) {
+        GST_ERROR_OBJECT(src, "Invalid caps");
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+static void webkit_flite_src_class_init(WebKitFliteSrcClass* klass)
+{
+    GObjectClass* objectClass = G_OBJECT_CLASS(klass);
+    objectClass->constructed = webkitFliteSrcConstructed;
+
+    GstElementClass* elementClass = GST_ELEMENT_CLASS(klass);
+    elementClass->change_state = webkitFliteSrcChangeState;
+
+    gst_element_class_add_static_pad_template(elementClass, &srcTemplate);
+
+    gst_element_class_set_static_metadata(elementClass,
+        "WebKit WebSpeech GstFlite source element", "Source",
+        "Handles WebSpeech data from WebCore",
+        "ChangSeok Oh <changseok@webkit.org>");
+
+    GstBaseSrcClass* baseSrcClass = GST_BASE_SRC_CLASS(klass);
+    baseSrcClass->start = GST_DEBUG_FUNCPTR(webkitFliteSrcStart);
+    baseSrcClass->stop = GST_DEBUG_FUNCPTR(webkitFliteSrcStop);
+    baseSrcClass->create = GST_DEBUG_FUNCPTR(webkitFliteSrcCreate);
+    baseSrcClass->set_caps = GST_DEBUG_FUNCPTR(webkitFliteSrcSetCaps);
+
+    flite_init();
+}
+
+Vector<RefPtr<PlatformSpeechSynthesisVoice>>& ensureFliteVoicesInitialized()
+{
+    // FIXME: Ugly hack. Some website does not call initializeVoiceList().
+    static Vector<RefPtr<PlatformSpeechSynthesisVoice>> voiceList;
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [] {
+        flite_add_lang("eng", usenglish_init, cmulex_init);
+        flite_add_lang("usenglish", usenglish_init, cmulex_init);
+
+        // The order of voices added does not hold. A newly added voice is inserted in the second place on the list.
+        flite_add_voice(register_cmu_us_kal(nullptr));
+        flite_add_voice(register_cmu_us_slt(nullptr));
+        flite_add_voice(register_cmu_us_rms(nullptr));
+        flite_add_voice(register_cmu_us_awb(nullptr));
+
+        for (const cst_val* v = flite_voice_list; v; v = val_cdr(v)) {
+            cst_voice* voice = val_voice(val_car(v));
+            String voiceName = String::fromUTF8(voice->name);
+            voiceList.append(PlatformSpeechSynthesisVoice::create("http://festvox.org/flite/packed/flite-2.1/voices/cmu_us_"_s + voiceName + ".flitevox"_s, voiceName, "en-US"_s, false, false));
```

Comment:
I wonder about other locales, can we load voices in other languages? is there something in the spec about that?

Reply:
Also, how is flite loading its "voices"? Is it using the network? If so, that could be a WebProcess sandbox violation.

Reply:
Regarding other local languages, Maybe yes. Available voices provided by the project owner are English and Indian. I will be back to you with more info about that.

We can select a voice in two ways through flite_voice_select(). One is giving a voice name, and the other is giving a URL. I agree with your security concern, so the URI string is removed.

Reply:
Also even if network usage was allowable, it would need to not use `http://`.

Reply:
@philn Correct. No network usage to get voices.
@mcatanzaro festvox.org does not support https, unfortunately. :(